### PR TITLE
Heimdall App implementation

### PR DESCRIPTION
**Pull request by 0xKrishna**
# Description

In this PR the goal is to implement heimdall interface using heimdallApp, So that we can directly communicate b/w bor & heimdall while they are running as a single process.

Some changes in this PR are also done to remove cyclic dependency, Basically some unused code which was creating that cyclic dependency is commented out.

Also the flags to run heimdall is marked as WIP so that people don't use it as of now, We will release this feature with proper doc after milestone HF.
# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [ ] I have added at least 2 reviewer or the whole pos-v1 team
- [ ] I have added sufficient documentation in code
- [ ] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai
- [ ] I have created new e2e tests into express-cli

### Manual tests

Please complete this section with the steps you performed if you ran manual tests for this functionality, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: consensus/bor/heimdallapp/state_sync.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package heimdallapp
+
+import (
+	"context"
+	"encoding/json"
```
Comment (JekaMas) [inline consensus/bor/heimdallapp/state_sync.go:5]: Should we use json-iter instead?

Comment (0xKrishna) [reply consensus/bor/heimdallapp/state_sync.go:5]: I am not sure, I used what I generally use to do json marshal/unmarshal, Also I checked other implementations of IHeimdallClient, We are using `encoding/json` there as well, Let me know if you want me to change/update.

**Review by temaniarpit27 — APPROVED** [review]

**Review by temaniarpit27 — APPROVED** [review]

File: consensus/bor/heimdallapp/checkpoint.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
+package heimdallapp
+
+import (
+	"context"
+	"encoding/json"
```
Comment (JekaMas) [inline consensus/bor/heimdallapp/checkpoint.go:5]: Could you use go-iterator instead?

Comment (0xKrishna) [reply consensus/bor/heimdallapp/checkpoint.go:5]: Okay

Comment (JekaMas) [reply consensus/bor/heimdallapp/checkpoint.go:5]: and in other places is the same

File: consensus/bor/heimdallapp/state_sync.go (hunk 1 above)
Comment (JekaMas) [reply consensus/bor/heimdallapp/state_sync.go:5]: check other places - we started using go-iterator for quite a while

File: consensus/bor/heimdallapp/checkpoint.go (hunk 2 above)
Comment (0xKrishna) [reply consensus/bor/heimdallapp/checkpoint.go:5]: Are you talking about this one ? https://pkg.go.dev/google.golang.org/api/iterator

Comment (0xKrishna) [reply consensus/bor/heimdallapp/checkpoint.go:5]: I think you were talking about this, My bad
https://pkg.go.dev/github.com/json-iterator/go

File: consensus/bor/heimdallapp/checkpoint.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	log.Info("Fetched checkpoint", "number", number)
+
+	resBytes, err := json.Marshal(res)
+	if err != nil {
+		return nil, err
+	}
+
+	var checkpoint checkpoint.Checkpoint
+
+	err = json.Unmarshal(resBytes, &checkpoint)
```
Comment (JekaMas) [inline consensus/bor/heimdallapp/checkpoint.go:44]: why we marshall to unmarshall then?

File: consensus/bor/heimdallapp/client.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+func NewHeimdallAppClient() *HeimdallAppClient {
+	return &HeimdallAppClient{
+		hApp: service.GetHeimdallApp(),
+	}
+}
+
+func (h *HeimdallAppClient) Close() {
+	// Nothing to close as of now
+	log.Debug("Shutdown detected, Closing Heimdall App conn")
```
Comment (JekaMas) [inline consensus/bor/heimdallapp/client.go:26]: maybe warn level logging?

File: consensus/bor/heimdallapp/checkpoint.go (hunk 2 above)
Comment (JekaMas) [reply consensus/bor/heimdallapp/checkpoint.go:5]: that one

**Review by JekaMas — APPROVED** [review]

File: consensus/bor/heimdallapp/state_sync.go (hunk 1 above)
Comment (0xKrishna) [reply consensus/bor/heimdallapp/state_sync.go:5]: Yes ser, On it

File: consensus/bor/heimdallapp/checkpoint.go (hunk 3 above)
Comment (0xKrishna) [reply consensus/bor/heimdallapp/checkpoint.go:44]: To convert the response from `types.Checkpoint` to `checkpoint.Checkpoint`
Basically they have similar data but one is from heimdall repo (from where we are fetching checkpoint) and other is bor repo's checkpoint object.

Comment (0xKrishna) [reply consensus/bor/heimdallapp/checkpoint.go:44]: We can't directly assign as the data and json tags are same, But their internall types are different.

https://github.com/maticnetwork/heimdall/blob/develop/types/header.go#L9

https://github.com/maticnetwork/bor/blob/develop/consensus/bor/heimdall/checkpoint/checkpoint.go#L10

File: consensus/bor/heimdallapp/span.go [hunk 5]
```diff
@@ -38,23 +37,10 @@ func toSpan(hdSpan *hmTypes.Span) *span.HeimdallSpan {
...
-	// set validator map
-	validatorsMap := make(map[common.Address]int, len(set.Validators))
-
-	for i, val := range set.Validators {
-		validatorsMap[val.Address] = i
-	}
-
-	set.SetMap(validatorsMap)
-
-	set.SetTotalVotingPower(vs.TotalVotingPower())
```
Comment (JekaMas) [inline consensus/bor/heimdallapp/span.go]: Don't we need to fill all fields?

Comment (0xKrishna) [reply consensus/bor/heimdallapp/span.go]: Nope, That is calculated in validator-set while processing. In other ways of communications (HTTP/gRPC) also we don't set totalVotingPower and the validator map as they are anyway calculated.

**Pull request merged**